core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -387,9 +388,12 @@ public long indexTranslogOperations(List<Translog.Operation> operations, int tot
             throw new IndexShardNotRecoveringException(shardId, indexShard().state());
         }
         for (Translog.Operation operation : operations) {
-            indexShard().applyTranslogOperation(operation, Engine.Operation.Origin.PEER_RECOVERY, update -> {
+            Engine.Result result = indexShard().applyTranslogOperation(operation, Engine.Operation.Origin.PEER_RECOVERY, update -> {
                 throw new MapperException("mapping updates are not allowed [" + operation + "]");
             });
+            assert result.hasFailure() == false ||
```

Comment:
as we discussed - corruptions can't lead to mapping parsing failures. They fail way earlier when reading from the translogs